File: api/src/lib/ssrf-guard.ts
/**
 * SSRF-Guard: Resolve-and-Pin + Block privater/link-local/loopback/metadata-Ranges.
 *
 * Härtung gegen DNS-Rebinding (VEC-175): Ein Angreifer, der die DNS seiner
 * Domain kontrolliert, kann nach bestandener Verifikation die Domain zur
 * Verbindungszeit auf eine interne IP (z. B. 169.254.169.254, RFC1918) zeigen.
 * Wir lösen genau einmal auf, validieren die Adresse gegen eine Blockliste und
 * verbinden gegen die so gepinnte öffentliche IP. Da `http.request({ lookup })`
 * die Auflösung selbst übernimmt, gibt es kein TOCTOU-Fenster zwischen unserer
 * Validierung und dem tatsächlichen Connect — der Connect nutzt exakt die von
 * uns freigegebene Adresse.
 */

import dns from 'dns';
import http from 'http';
import https from 'https';
import net from 'net';
import { URL } from 'url';

export const DEFAULT_TIMEOUT_MS = 10_000;
export const DEFAULT_MAX_REDIRECTS = 5;
export const DEFAULT_MAX_BYTES = 1_000_000; // 1 MB — reicht für TXT/HTML-Verify

/** Fehler, der bei geblockter (interner) Zieladresse geworfen wird. */
export class SsrfBlockedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SsrfBlockedError';
  }
}

function ipv4ToInt(ip: string): number | null {
  const parts = ip.split('.');
  if (parts.length !== 4) return null;
  let value = 0;
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null;
    const n = Number(part);
    if (n > 255) return null;
    value = value * 256 + n;
  }
  return value >>> 0;
}

function inV4Cidr(ipInt: number, network: string, prefix: number): boolean {
  const netInt = ipv4ToInt(network);
  if (netInt === null) return false;
  if (prefix === 0) return true;
  const mask = (0xffffffff << (32 - prefix)) >>> 0;
  return (ipInt & mask) === (netInt & mask);
}

// RFC1918 + Loopback + Link-Local (inkl. 169.254.169.254 Cloud-Metadata) +
// CGNAT, Benchmark, "this network", Multicast, reserviert/Broadcast.
const BLOCKED_V4: Array<[string, number]> = [
  ['0.0.0.0', 8],
  ['10.0.0.0', 8],
  ['100.64.0.0', 10],
  ['127.0.0.0', 8],
  ['169.254.0.0', 16],
  ['172.16.0.0', 12],
  ['192.0.0.0', 24],
  ['192.0.2.0', 24],
  ['192.168.0.0', 16],
  ['198.18.0.0', 15],
  ['198.51.100.0', 24],
  ['203.0.113.0', 24],
  ['224.0.0.0', 4],
  ['240.0.0.0', 4],
  ['255.255.255.255', 32],
];

function isBlockedV4(ip: string): boolean {
  const ipInt = ipv4ToInt(ip);
  if (ipInt === null) return true; // unparsebar → fail-closed
  return BLOCKED_V4.some(([net, prefix]) => inV4Cidr(ipInt, net, prefix));
}

/** Expandiert einen IPv6-String auf 16 Bytes oder gibt null bei Fehler. */
function ipv6ToBytes(ip: string): Uint8Array | null {
  let zone = ip;
  const pct = zone.indexOf('%');
  if (pct !== -1) zone = zone.slice(0, pct); // Scope-ID entfernen
  if (zone.indexOf(':') === -1) return null;

  const halves = zone.split('::');
  if (halves.length > 2) return null;

  const parseGroups = (segment: string): number[] | null => {
    if (segment === '') return [];
    const groups: number[] = [];
    for (const g of segment.split(':')) {
      if (g.indexOf('.') !== -1) {
        // Eingebettete IPv4 (z. B. ::ffff:1.2.3.4)
        const v4 = ipv4ToInt(g);
        if (v4 === null) return null;
        groups.push((v4 >>> 16) & 0xffff, v4 & 0xffff);
      } else {
        if (!/^[0-9a-fA-F]{1,4}$/.test(g)) return null;
        groups.push(parseInt(g, 16));
      }
    }
    return groups;
  };

  const head = parseGroups(halves[0]);
  const tail = halves.length === 2 ? parseGroups(halves[1]) : [];
  if (head === null || tail === null) return null;

  let groups: number[];
  if (halves.length === 2) {
    const fill = 8 - head.length - tail.length;
    if (fill < 0) return null;
    groups = [...head, ...new Array(fill).fill(0), ...tail];
  } else {
    groups = head;
  }
  if (groups.length !== 8) return null;

  const bytes = new Uint8Array(16);
  for (let i = 0; i < 8; i++) {
    bytes[i * 2] = (groups[i] >> 8) & 0xff;
    bytes[i * 2 + 1] = groups[i] & 0xff;
  }
  return bytes;
}

function isBlockedV6(ip: string): boolean {
  const b = ipv6ToBytes(ip);
  if (b === null) return true; // fail-closed

  // ::  (unspecified) und ::1 (loopback)
  const allButLastZero = b.slice(0, 15).every((x) => x === 0);
  if (allButLastZero && (b[15] === 0 || b[15] === 1)) return true;

  // fc00::/7  ULA  (erstes Byte 0xfc oder 0xfd)
  if ((b[0] & 0xfe) === 0xfc) return true;

  // fe80::/10  link-local  und  fec0::/10  site-local (deprecated)
  if (b[0] === 0xfe && (b[1] & 0xc0) === 0x80) return true;
  if (b[0] === 0xfe && (b[1] & 0xc0) === 0xc0) return true;

  // ff00::/8  multicast
  if (b[0] === 0xff) return true;

  // ::ffff:0:0/96  IPv4-mapped → eingebettete IPv4 prüfen
  const mappedPrefix =
    b.slice(0, 10).every((x) => x === 0) && b[10] === 0xff && b[11] === 0xff;
  if (mappedPrefix) {
    const v4 = `${b[12]}.${b[13]}.${b[14]}.${b[15]}`;
    return isBlockedV4(v4);
  }

  // 64:ff9b::/96  NAT64 → eingebettete IPv4 prüfen
  if (b[0] === 0x00 && b[1] === 0x64 && b[2] === 0xff && b[3] === 0x9b) {
    const v4 = `${b[12]}.${b[13]}.${b[14]}.${b[15]}`;
    return isBlockedV4(v4);
  }

  return false;
}

/**
 * True, wenn die Adresse in einem geblockten (internen/reservierten) Bereich
 * liegt. Unparsebare Eingaben gelten als geblockt (fail-closed).
 */
export function isBlockedAddress(ip: string): boolean {
  const family = net.isIP(ip);
  if (family === 4) return isBlockedV4(ip);
  if (family === 6) return isBlockedV6(ip);
  return true; // kein gültiges IP-Literal → fail-closed
}

/**
 * `lookup`-Funktion kompatibel mit http(s).request: löst den Hostnamen auf,
 * verwirft geblockte Adressen und pinnt auf die erste verbleibende öffentliche
 * IP. Bleibt keine öffentliche Adresse übrig → Fehler (fail-closed).
 *
 * `resolver` ist injizierbar für Tests; Default ist dns.lookup(all).
 */
export function makeGuardedLookup(
  resolver: (
    hostname: string,
    options: { all: true },
    cb: (err: NodeJS.ErrnoException | null, addresses: dns.LookupAddress[]) => void,
  ) => void = dns.lookup as never,
) {
  return function guardedLookup(
    hostname: string,
    options: dns.LookupOneOptions | dns.LookupAllOptions | number,
    callback: (
      err: NodeJS.ErrnoException | null,
      address: string | dns.LookupAddress[],
      family?: number,
    ) => void,
  ): void {
    // Node ruft `lookup` je nach Aufrufer mit `all: true` (z. B. bei
    // autoSelectFamily/Happy-Eyeballs, Default in Node ≥18) ODER `all: false`
    // auf. Im all-Modus MUSS ein Array zurückgegeben werden, sonst wirft Node
    // ERR_INVALID_IP_ADDRESS und die Verbindung schlägt für JEDEN legitimen
    // Host fehl. Wir respektieren den Modus des Aufrufers.
    const wantAll =
      typeof options === 'object' && options !== null && (options as dns.LookupAllOptions).all === true;
    const succeed = (address: string, family: number): void => {
      if (wantAll) {
        callback(null, [{ address, family }], family);
      } else {
        callback(null, address, family);
      }
    };

    // IP-Literale direkt prüfen (keine Auflösung nötig).
    const literalFamily = net.isIP(hostname);
    if (literalFamily !== 0) {
      if (isBlockedAddress(hostname)) {
        callback(new SsrfBlockedError(`Geblockte Zieladresse: ${hostname}`), '', literalFamily);
        return;
      }
      succeed(hostname, literalFamily);
      return;
    }

    resolver(hostname, { all: true }, (err, addresses) => {
      if (err) {
        callback(err, '', 0);
        return;
      }
      const allowed = addresses.filter((a) => !isBlockedAddress(a.address));
      if (allowed.length === 0) {
        callback(
          new SsrfBlockedError(
            `Alle aufgelösten Adressen für ${hostname} sind geblockt (mögliches DNS-Rebinding).`,
          ),
          '',
          0,
        );
        return;
      }
      const pinned = allowed[0];
      succeed(pinned.address, pinned.family);
    });
  };
}

const guardedLookup = makeGuardedLookup();

export interface SafeResponse {
  status: number;
  url: string;
  text(): Promise<string>;
}

export interface SafeFetchOptions {
  timeoutMs?: number;
  maxRedirects?: number;
  maxBytes?: number;
  headers?: Record<string, string>;
}

/**
 * Minimaler HTTP(S)-GET mit SSRF-Schutz: pinnt jede Verbindung (auch nach
 * Redirects) auf eine validierte öffentliche IP, folgt Redirects gebounded und
 * begrenzt die Body-Größe. Nur http/https-Schemata sind erlaubt.
 */
export function safeFetch(url: string, opts: SafeFetchOptions = {}): Promise<SafeResponse> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const maxRedirects = opts.maxRedirects ?? DEFAULT_MAX_REDIRECTS;
  const maxBytes = opts.maxBytes ?? DEFAULT_MAX_BYTES;

  return new Promise<SafeResponse>((resolve, reject) => {
    const visit = (current: string, redirectsLeft: number): void => {
      let parsed: URL;
      try {
        parsed = new URL(current);
      } catch {
        reject(new SsrfBlockedError(`Ungültige URL: ${current}`));
        return;
      }
      if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') {
        reject(new SsrfBlockedError(`Nicht erlaubtes Schema: ${parsed.protocol}`));
        return;
      }

      // IP-Literal-Hosts überspringt Node's net.connect bei der DNS-Auflösung,
      // d. h. unser `lookup`-Guard würde NICHT aufgerufen. Daher hier explizit
      // prüfen (URL-Hostname trägt IPv6 in eckigen Klammern → strippen).
      const rawHost = parsed.hostname.replace(/^\[|\]$/g, '');
      if (net.isIP(rawHost) !== 0 && isBlockedAddress(rawHost)) {
        reject(new SsrfBlockedError(`Geblockte Ziel-IP: ${rawHost}`));
        return;
      }

      const transport = parsed.protocol === 'https:' ? https : http;
      const req = transport.request(
        current,
        {
          method: 'GET',
          lookup: guardedLookup,
          headers: { 'User-Agent': 'VectiScan-Verify/1.0', ...(opts.headers ?? {}) },
        },
        (res) => {
          const status = res.statusCode ?? 0;
          const location = res.headers.location;
          if (status >= 300 && status < 400 && location) {
            res.resume(); // Body verwerfen
            if (redirectsLeft <= 0) {
              reject(new SsrfBlockedError('Zu viele Redirects.'));
              return;
            }
            let next: string;
            try {
              next = new URL(location, current).toString();
            } catch {
              reject(new SsrfBlockedError(`Ungültiger Redirect: ${location}`));
              return;
            }
            visit(next, redirectsLeft - 1);
            return;
          }

          const chunks: Buffer[] = [];
          let total = 0;
          res.on('data', (chunk: Buffer) => {
            total += chunk.length;
            if (total > maxBytes) {
              req.destroy();
              reject(new SsrfBlockedError('Antwort überschreitet Größenlimit.'));
              return;
            }
            chunks.push(chunk);
          });
          res.on('end', () => {
            const body = Buffer.concat(chunks).toString('utf8');
            resolve({ status, url: current, text: () => Promise.resolve(body) });
          });
        },
      );

      req.setTimeout(timeoutMs, () => {
        req.destroy(new Error('Timeout'));
      });
      req.on('error', (err) => reject(err));
      req.end();
    };

    visit(url, maxRedirects);
  });
}

File: scan-worker/scanner/common/ssrf_guard.py
"""SSRF-Guard fuer den Python-Scan-Worker (VEC-196, Follow-up aus VEC-175).

Spiegelt die Semantik von ``api/src/lib/ssrf-guard.ts`` im Python-Worker:

1. Zentraler, **fail-closed** IP-Block-Helfer fuer geblockte/interne Bereiche:
   loopback, RFC1918, link-local (inkl. 169.254.169.254 Cloud-Metadata),
   CGNAT (100.64/10), Benchmark (198.18/15), "this network", reserviert,
   Multicast, Broadcast sowie IPv6 ULA/link-local/site-local/multicast und die
   eingebetteten IPv4-Faelle IPv4-mapped (``::ffff:0:0/96``) und NAT64
   (``64:ff9b::/96``).

2. **Resolve-and-Pin** fuer ``requests``: Der Hostname wird genau einmal
   aufgeloest, geblockte Adressen werden verworfen und die Verbindung gegen die
   gepinnte oeffentliche IP-Literal-URL aufgebaut. Weil der Connect gegen ein
   IP-Literal geht, findet keine erneute DNS-Aufloesung statt -> es gibt kein
   TOCTOU-/DNS-Rebinding-Fenster zwischen Validierung und Verbindung. Die
   Validierung laeuft pro ``send()`` und damit auch pro Redirect-Hop.

Die explizite IPv4-Blockliste ist 1:1 an ``ssrf-guard.ts`` gespiegelt und
verlaesst sich bewusst NICHT allein auf ``ipaddress.is_private`` (dessen
Definition zwischen Python-Versionen variiert), sondern auf
CIDR-Zugehoerigkeit -> deterministisch und version-unabhaengig.
"""

from __future__ import annotations

import ipaddress
from typing import Iterable, Optional
from urllib.parse import urlparse, urlunparse

import requests
from requests.adapters import HTTPAdapter


DEFAULT_RESOLVE_TIMEOUT = 5.0
DEFAULT_MAX_REDIRECTS = 5


class SsrfBlockedError(Exception):
    """Zieladresse liegt in einem geblockten (internen) Bereich (fail-closed)."""


# Explizite, an ssrf-guard.ts gespiegelte IPv4-Blockliste.
_BLOCKED_V4 = [
    ipaddress.ip_network(cidr)
    for cidr in (
        "0.0.0.0/8",          # "this network"
        "10.0.0.0/8",         # RFC1918
        "100.64.0.0/10",      # CGNAT
        "127.0.0.0/8",        # loopback
        "169.254.0.0/16",     # link-local (inkl. 169.254.169.254 Metadata)
        "172.16.0.0/12",      # RFC1918
        "192.0.0.0/24",       # IETF protocol assignments
        "192.0.2.0/24",       # TEST-NET-1
        "192.168.0.0/16",     # RFC1918
        "198.18.0.0/15",      # benchmark
        "198.51.100.0/24",    # TEST-NET-2
        "203.0.113.0/24",     # TEST-NET-3
        "224.0.0.0/4",        # multicast
        "240.0.0.0/4",        # reserviert
        "255.255.255.255/32", # broadcast
    )
]

_NAT64_NET = ipaddress.ip_network("64:ff9b::/96")


def _is_blocked_v4(addr: ipaddress.IPv4Address) -> bool:
    return any(addr in net for net in _BLOCKED_V4)


def _is_blocked_v6(addr: ipaddress.IPv6Address) -> bool:
    # IPv4-mapped (::ffff:0:0/96) -> eingebettete IPv4 pruefen.
    mapped = addr.ipv4_mapped
    if mapped is not None:
        return _is_blocked_v4(mapped)
    # NAT64 64:ff9b::/96 -> letzte 32 Bit als IPv4 pruefen.
    if addr in _NAT64_NET:
        return _is_blocked_v4(ipaddress.IPv4Address(int(addr) & 0xFFFFFFFF))
    # loopback (::1), link-local (fe80::/10), multicast (ff00::/8),
    # ULA (fc00::/7) zaehlt als is_private, unspecified (::),
    # site-local (fec0::/10, deprecated), reserviert.
    return bool(
        addr.is_loopback
        or addr.is_link_local
        or addr.is_multicast
        or addr.is_private
        or addr.is_unspecified
        or addr.is_site_local
        or addr.is_reserved
    )


def is_blocked_address(ip: str) -> bool:
    """True, wenn ``ip`` in einem geblockten Bereich liegt.

    Unparsebare Eingaben gelten als geblockt (fail-closed).
    """
    try:
        addr = ipaddress.ip_address(ip.strip())
    except (ValueError, AttributeError):
        return True
    if isinstance(addr, ipaddress.IPv4Address):
        return _is_blocked_v4(addr)
    return _is_blocked_v6(addr)


def filter_public(ips: Iterable[str]) -> list[str]:
    """Behaelt nur oeffentliche (nicht geblockte) Adressen, Reihenfolge stabil."""
    return [ip for ip in ips if not is_blocked_address(ip)]


def _resolve_host(host: str, timeout: float) -> list[str]:
    """A + AAAA ueber die fixen Resolver (dns_utils). Lazy-Import gegen Zyklus."""
    from scanner.common import dns_utils

    return list(dns_utils.resolve_a(host, timeout)) + list(
        dns_utils.resolve_aaaa(host, timeout)
    )


def resolve_and_pin(host: str, timeout: float = DEFAULT_RESOLVE_TIMEOUT) -> str:
    """Loese ``host`` auf, verwirf geblockte IPs, pinne erste oeffentliche IP.

    Ist ``host`` bereits ein IP-Literal, wird nur die Blockliste geprueft.
    Raises ``SsrfBlockedError``, wenn ``host`` geblockt ist, nicht aufloesbar
    ist oder nur interne Adressen uebrig bleiben (moegliches DNS-Rebinding).
    """
    try:
        ipaddress.ip_address(host)
        is_literal = True
    except ValueError:
        is_literal = False

    if is_literal:
        if is_blocked_address(host):
            raise SsrfBlockedError(f"Geblockte Ziel-IP: {host}")
        return host

    resolved = _resolve_host(host, timeout)
    if not resolved:
        raise SsrfBlockedError(f"Keine DNS-Aufloesung fuer {host}")
    public = filter_public(resolved)
    if not public:
        raise SsrfBlockedError(
            f"Alle aufgeloesten Adressen fuer {host} sind geblockt "
            f"(moegliches DNS-Rebinding)."
        )
    return public[0]


def _host_for_url(ip: str) -> str:
    return f"[{ip}]" if ":" in ip else ip


class PinnedHTTPAdapter(HTTPAdapter):
    """``requests``-Adapter mit Resolve-and-Pin.

    Bei jedem ``send()`` (auch pro Redirect-Hop) wird der Ziel-Host aufgeloest,
    gegen die Blockliste validiert und die Verbindung auf die gepinnte
    oeffentliche IP umgeschrieben. Der urspruengliche Hostname wandert in den
    ``Host``-Header, damit vhost-basierte Server korrekt antworten.

    Tech-Debt (low, VEC-196): Bei HTTPS geht die SNI gegen das IP-Literal, nicht
    gegen den Hostnamen. Fuer den Probe-Pfad (``verify=False``) unkritisch; eine
    SNI-erhaltende Variante (ForcedIP-Adapter) ist als Folge-Optimierung notiert.
    """

    def __init__(self, *args, resolve_timeout: float = DEFAULT_RESOLVE_TIMEOUT, **kwargs):
        self._resolve_timeout = resolve_timeout
        super().__init__(*args, **kwargs)

    def send(self, request, **kwargs):  # type: ignore[override]
        parsed = urlparse(request.url)
        host = parsed.hostname
        if not host:
            raise SsrfBlockedError(f"URL ohne Host: {request.url}")

        pinned = resolve_and_pin(host, self._resolve_timeout)
        if pinned != host:
            host_header = host if not parsed.port else f"{host}:{parsed.port}"
            request.headers["Host"] = host_header
            netloc = _host_for_url(pinned)
            if parsed.port:
                netloc = f"{netloc}:{parsed.port}"
            request.url = urlunparse(parsed._replace(netloc=netloc))
        return super().send(request, **kwargs)


def guarded_session(
    verify: bool = False, resolve_timeout: float = DEFAULT_RESOLVE_TIMEOUT
) -> requests.Session:
    """Eine ``requests.Session`` mit gemountetem :class:`PinnedHTTPAdapter`."""
    session = requests.Session()
    adapter = PinnedHTTPAdapter(resolve_timeout=resolve_timeout)
    session.mount("http://", adapter)
    session.mount("https://", adapter)
    session.verify = verify
    return session


def safe_get(
    url: str,
    *,
    timeout: float = DEFAULT_RESOLVE_TIMEOUT,
    allow_redirects: bool = True,
    headers: Optional[dict] = None,
    verify: bool = False,
    max_redirects: int = DEFAULT_MAX_REDIRECTS,
) -> requests.Response:
    """SSRF-gepinntes ``GET``. Wirft :class:`SsrfBlockedError` bei internem Ziel."""
    session = guarded_session(verify=verify, resolve_timeout=timeout)
    session.max_redirects = max_redirects
    return session.get(
        url, timeout=timeout, allow_redirects=allow_redirects, headers=headers or {}
    )
